plugins/inputs/tencentcloudcm/discover.go

```diff
@@ -0,0 +1,213 @@
+func (d *discoverTool) GetInstance(account, namespace, region, key string) map[string]interface{} {
+	d.rw.RLock()
+	defer d.rw.RUnlock()
+	discoverObject, ok := d.DiscoveredObjects[newKey(account, namespace, region)]
+	if !ok {
+		return nil
+	}
+	instance, ok := discoverObject.Instances[key]
+	if !ok {
+		return nil
+	}
+	return instance
+}
+
+// GetMonitorInstances Get discovered monitor instances
+func (d *discoverTool) GetMonitorInstances(account, namespace, region string) []*monitor.Instance {
+	d.rw.RLock()
+	defer d.rw.RUnlock()
+	discoverObject, ok := d.DiscoveredObjects[newKey(account, namespace, region)]
+	if !ok {
+		return nil
+	}
+	return discoverObject.MonitorInstances
+}
+
+// Discover implements Product interface
+func Discover(crs *common.Credential, region, endpoint string, p Product) (discoverObject, error) {
+	discoverObject := discoverObject{
+		Instances: map[string]map[string]interface{}{},
+	}
+
+	offset, limit := int64(0), int64(100)
+	// dcInstances := []*dc.DirectConnectTunnel{}
+	instances := []map[string]interface{}{}
+
+	total, instancesByPage, err := p.Discover(crs, region, endpoint, offset, limit)
+	if err != nil {
+		return discoverObject, err
+	}
+	instances = append(instances, instancesByPage...)
+
+	// discover all instances if total count is bigger than limit
+	for i := 1; i < int(int64(total)/limit)+1; i++ {
+		_, instancesByPage, err := p.Discover(crs, region, endpoint, offset+(int64(i)*limit), limit)
+		if err != nil {
+			return discoverObject, err
+		}
+		instances = append(instances, instancesByPage...)
+	}
+
+	monitorInstances := []*monitor.Instance{}
+	for _, instance := range instances {
+		keyIsNil := false
+		keyVals := []interface{}{}
+		dimensions := []*monitor.Dimension{}
+
+		// normalized instance have lower case field name
+		normInstance := map[string]interface{}{}
+		for k, v := range instance {
+			normInstance[strings.ToLower(k)] = v
+		}
+
+		for _, key := range p.Keys() {
+			// check if discovered key field is nil
+			if normInstance[strings.ToLower(key)] == nil {
+				keyIsNil = true
+				break
+			} else {
+				// construct keyVals and dimensions based on the key and discovered instance
+				keyVals = append(keyVals, normInstance[strings.ToLower(key)])
+				dimensions = append(dimensions, &monitor.Dimension{
+					Name:  common.StringPtr(key),
+					Value: common.StringPtr(fmt.Sprintf("%v", normInstance[strings.ToLower(key)])),
+				})
+			}
+		}
+
+		// instance with nil key field will be discarded
+		if keyIsNil {
+			continue
+		}
+
+		discoverObject.Instances[newKey(keyVals...)] = instance
+		monitorInstances = append(monitorInstances, &monitor.Instance{Dimensions: dimensions})
+	}
+	discoverObject.MonitorInstances = monitorInstances
+
+	return discoverObject, nil
+}
+
+func newKey(vals ...interface{}) string {
+	key := ""
+	for i, val := range vals {
+		if i != len(vals)-1 {
+			key += fmt.Sprintf("%s-", val)
+		} else {
+			key += fmt.Sprintf("%s", val)
+		}
+	}
+	return key
```

Comment:
How about `strings.Join(vals, "-")`? See https://pkg.go.dev/strings#Join...

Reply:
fixed in commit 114c8bd